README.md
[![Supported Versions](https://img.shields.io/pypi/pyversions/scrapli_cfg.svg)](https://pypi.org/project/scrapli_cfg)
[![PyPI version](https://badge.fury.io/py/scrapli-cfg.svg)](https://badge.fury.io/py/scrapli-cfg)
[![Weekly Build](https://github.com/scrapli/scrapli_cfg/workflows/Weekly%20Build/badge.svg)](https://github.com/scrapli/scrapli_cfg/actions?query=workflow%3A%22Weekly+Build%22)
[![Code Style](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/ambv/black)
[![License: MIT](https://img.shields.io/badge/License-MIT-blueviolet.svg)](https://opensource.org/licenses/MIT)

scrapli_cfg
===========

---

**Documentation**: <a href="https://scrapli.github.io/scrapli_cfg" target="_blank">https://scrapli.github.io/scrapli_cfg</a>

**Source Code**: <a href="https://github.com/scrapli/scrapli_cfg" target="_blank">https://github.com/scrapli/scrapli_cfg</a>

**Examples**: <a href="https://github.com/scrapli/scrapli_cfg/tree/master/examples" target="_blank">https://github.com/scrapli/scrapli_cfg/tree/master/examples</a>

---

scrapli_cfg makes merging or replacing device configurations over Telnet or SSH easy, all while giving you the 
scrapli behaviour you know and love.


#### Key Features:

- __Easy__: It's easy to get going with scrapli and scrapli-cfg -- check out the documentation and example links above, 
  and you'll be managing device configurations in no time.
- __Fast__: Do you like to go fast? Of course you do! All of scrapli is built with speed in mind, but if you really 
  feel the need for speed, check out the `ssh2` transport plugin to take it to the next level! All the "normal" 
  scrapli transport plugin goodness exists here in scrapli-cfg too!
- __Great Developer Experience__: scrapli_cfg has great editor support thanks to being fully typed; that plus 
  thorough docs make developing with scrapli a breeze.


## Requirements

MacOS or \*nix<sup>1</sup>, Python 3.7+

scrapli_cfg's only requirements is `scrapli`.

<sup>1</sup> Although many parts of scrapli *do* run on Windows, Windows is not officially supported


## Installation

```
pip install scrapli-cfg
```

See the [docs](https://scrapli.github.io/scrapli_cfg/user_guide/installation) for other installation methods/details.



## A simple Example

```python
from scrapli import Scrapli
from scrapli_cfg import ScrapliCfg

device = {
   "host": "172.18.0.11",
   "auth_username": "scrapli",
   "auth_password": "scrapli",
   "auth_strict_key": False,
   "platform": "cisco_iosxe"
}

with open("myconfig", "r") as f:
    my_config = f.read()

with Scrapli(**device) as conn:
  cfg_conn = ScrapliCfg(conn=conn)
  cfg_conn.prepare()
  cfg_conn.load_config(config=my_config, replace=True)
  diff = cfg_conn.diff_config()
  print(diff.side_by_side_diff)
  cfg_conn.commit_config()
  cfg_conn.cleanup()

```


docs/changelog.md
Changelog
=========

## 2022.07.30

- Fix from @WillIrvine to sort a very bad (carl's fault) regex overly aggressive matching -- see #41


## 2022.01.30

- Revised juniper abort config to remove candidate config file *after* rollback 0 to avoid issues where junos would 
  prompt for confirmation when exiting config mode to go delete the candidate config file prompting timeouts.
- Dropped Python3.6 support as it is now EOL! Of course, scrapli probably still works just fine with 3.6 (if you 
  install the old 3.6 requirements), but we won't test/support it anymore.
- Wow, pretty empty here... guess that's a good sign things have been working :p


## 2021.07.30

- Initial release!


noxfile.py
"""scrapli_cfg.noxfile"""
import re
import sys
from pathlib import Path
from typing import Dict, List

import nox

nox.options.error_on_missing_interpreters = False
nox.options.stop_on_first_error = False
nox.options.default_venv_backend = "venv"


def parse_requirements(dev: bool = True) -> Dict[str, str]:
    """
    Parse requirements file

    Args:
        dev: parse dev requirements (or not)

    Returns:
        dict: dict of parsed requirements

    Raises:
        N/A

    """
    requirements = {}
    requirements_file = "requirements.txt" if not dev else "requirements-dev.txt"

    with open(requirements_file, "r", encoding="utf-8") as f:
        requirements_file_lines = f.readlines()

    requirements_lines: List[str] = [
        line
        for line in requirements_file_lines
        if not line.startswith("-r") and not line.startswith("#") and not line.startswith("-e")
    ]
    editable_requirements_lines: List[str] = [
        line for line in requirements_file_lines if line.startswith("-e")
    ]

    for requirement in requirements_lines:
        parsed_requirement = re.match(
            pattern=r"^([a-z0-9\-\_\.]+)([><=]{1,2}\S*)(?:.*)$",
            string=requirement,
            flags=re.I | re.M,
        )
        requirements[parsed_requirement.groups()[0]] = parsed_requirement.groups()[1]

    for requirement in editable_requirements_lines:
        parsed_requirement = re.match(
            pattern=r"^-e\s.*(?:#egg=)(\w+)$", string=requirement, flags=re.I | re.M
        )
        requirements[parsed_requirement.groups()[0]] = requirement

    return requirements


REQUIREMENTS: Dict[str, str] = parse_requirements(dev=False)
DEV_REQUIREMENTS: Dict[str, str] = parse_requirements(dev=True)
PLATFORM: str = sys.platform
SKIP_LIST: List[str] = [
    "unit_tests-darwin-3.11",
    "integration_tests-darwin-3.11",
]


@nox.session(python=["3.7", "3.8", "3.9", "3.10", "3.11"])
def unit_tests(session):
    """
    Nox run unit tests

    Args:
        session: nox session

    Returns:
        None

    Raises:
        N/A

    """
    if f"unit_tests-{PLATFORM}-{session.python}" in SKIP_LIST:
        return

    session.install("-r", "requirements-dev.txt")
    session.install(".")
    session.run(
        "python",
        "-m",
        "pytest",
        "--cov=scrapli_cfg",
        "--cov-report",
        "xml",
        "--cov-report",
        "term",
        "tests/unit",
        "-v",
    )


@nox.session(python=["3.10"])
def integration_tests(session):
    """
    Nox run integration tests

    Args:
        session: nox session

    Returns:
        None

    Raises:
        N/A

    """
    if f"integration_tests-{PLATFORM}-{session.python}" in SKIP_LIST:
        return

    session.install("-r", "requirements-dev.txt")
    session.install(".")
    # setting scrapli boxen -> 1 so that the saved scrapli replay sessions are "correctly"
    # pointing to the boxen dev env (i.e. port 21022 instead of 22 for iosxe, etc.)
    session.run(
        "python",
        "-m",
        "pytest",
        "--cov=scrapli_cfg",
        "--cov-report",
        "xml",
        "--cov-report",
        "term",
        "tests/integration",
        "-v",
        env={"SCRAPLI_BOXEN": "1"},
    )


@nox.session(python=["3.10"])
def isort(session):
    """
    Nox run isort

    Args:
        session: nox session

    Returns:
        None

    Raises:
        N/A

    """
    session.install(f"isort{DEV_REQUIREMENTS['isort']}")
    session.run("python", "-m", "isort", "-c", ".")


@nox.session(python=["3.10"])
def black(session):
    """
    Nox run black

    Args:
        session: nox session

    Returns:
        None

    Raises:
        N/A

    """
    session.install(f"black{DEV_REQUIREMENTS['black']}")
    session.run("python", "-m", "black", "--check", ".")


@nox.session(python=["3.10"])
def pylama(session):
    """
    Nox run pylama

    Args:
        session: nox session

    Returns:
        None

    Raises:
        N/A

    """
    session.install("-r", "requirements-dev.txt")
    session.run("python", "-m", "pylama", ".")


@nox.session(python=["3.10"])
def pydocstyle(session):
    """
    Nox run pydocstyle

    Args:
        session: nox session

    Returns:
        None

    Raises:
        N/A

    """
    session.install(f"pydocstyle{DEV_REQUIREMENTS['pydocstyle']}")
    session.run("python", "-m", "pydocstyle", ".")


@nox.session(python=["3.10"])
def mypy(session):
    """
    Nox run mypy

    Args:
        session: nox session

    Returns:
        None

    Raises:
        N/A

    """
    session.install(".")
    session.install(f"mypy{DEV_REQUIREMENTS['mypy']}")
    session.run("python", "-m", "mypy", "--strict", "scrapli_cfg/")


@nox.session(python=["3.10"])
def darglint(session):
    """
    Nox run darglint

    Args:
        session: nox session

    Returns:
        None

    Raises:
        N/A

    """
    session.install(f"darglint{DEV_REQUIREMENTS['darglint']}")
    for file in Path("scrapli_cfg").rglob("*.py"):
        session.run("darglint", f"{file.absolute()}")
